Add web audio profiling article (#2757)
GoogleChrome/web.dev: src/site/content/en/blog/profiling-web-audio-apps-in-chrome/index.md

The screenshot below ([crbug.com/796330](https://crbug.com/796330)) is a
typical example of when code in `AudioWorkletProcessor` takes too long and
goes beyond a given render budget. The callback timing is well behaved but
the audio processing function call of Web Audio API failed to complete the
work before the next device callback.

<img src="5e-render-budget.png" alt="TODO" class="w-screenshot w-screenshot-filled">

**Your options:**

* Reduce the workload of the audio graph. (Use fewer `AudioNodes`.)
* Reduce the workload of your code in the `AudioWorkletProcessor`.
* Increase the base latency of `AudioContext`.

#### Example 2: Significant garbage collection on the worklet thread

Unlike on the native audio rendering thread, garbage collection is managed
on the worklet thread. That means if your code does memory allocation/deallocation
(e.g. new arrays) it eventually triggers a garbage collection which
synchronously blocks the thread. If the workload of web audio operations and
garbage collection is bigger than a given render budget, it results in
glitches in the audio stream. The screenshot is an extreme example of this
case.

> <img src="https://via.placeholder.com/250" alt="TODO" class="w-screenshot w-screenshot-filled">

Author: @hoch - missing image.

    - [Audio Device Client Prototype][aud-dev-client-proto]

#### Example 3: Jittery audio device callback from AudioOutputDevice

The precise timing of audio callback is the most important thing for web audio.
This should be the most precise clock in your system. If the operating system
or its audio subsystem cannot guarantee a solid callback timing, all the
subsequent operations will be impacted. The image shown below is an example
of jittery audio callback. Compared to the previous two images, the interval
between each callback varies significantly.

<img src="4e-pulse-audio.png" alt="TODO" class="w-screenshot w-screenshot-filled">

This is a known issue of Linux platform, which uses Pulse Audio as an audio
backend. This is still under investigation. (See
[crbug.com/825823](https://crbug.com/825823).)

## Use WebAudio DevTools

You also can use the DevTools panel specifically designed for web audio. This
is less comprehensive compared to the tracing tool, but it is useful if you
want to gauge the running performance of your application.

Access the panel by clicking the "three dots" menu in the top right corner of
DevTools, then go to **More tools** then **WebAudio**.

> <img src="https://via.placeholder.com/250" alt="TODO" class="w-screenshot w-screenshot-filled">

Author: @hoch - missing image.

### WebAudio Panel

> <img src="https://via.placeholder.com/250" alt="TODO" class="w-screenshot w-screenshot-filled">

Author: @hoch - missing image.

---
title: Profiling Web Audio in Chrome
subhead: Learn how to profile the performance of Web Audio apps in Chrome using chrome://tracing and the WebAudio panel in DevTools.
description: Learn how to profile the performance of Web Audio apps in Chrome using chrome://tracing and the WebAudio panel in DevTools.
date: 2020-04-30
updated: 2020-04-30
tags:
  - post # post is a required tag for the article to show up in the blog.
  - media
  - webaudio
authors:

>   - hongchanchoi

Reviewer: We should add a hero image or else the post won't show up in the **Latest Stories** section of the homepage.

https://web.dev/handbook/markup-media/#hero

We get most of our hero images from unsplash.com

Author: added

---

You reached this documentation probably because you're developing on an app
that uses the Web Audio API and experienced some unexpected glitches such as
popping noises from the output. You might already be involved in a
[crbug.com](https://crbug.com) discussion and a Chrome engineer asked you to
upload "tracing data". This document shows how to obtain the data so you can
help engineers triage, and eventually fix the issue.

## Web Audio profiling tools

There are two tools that will help you when profiling Web Audio,

> `chrome://tracing` and the WebAudio panel in Chrome DevTools.

Reviewer: ```suggestion
`chrome://tracing` and the **WebAudio** tab in Chrome DevTools.
```
* Formatted bold per style guide: https://developers.google.com/style/ui-elements#formatting
* Renamed to `tab` to keep it consistent with the official Chrome DevTools documentation (which I wrote most of)

> ### When do you chrome://tracing?

Reviewer: ```suggestion
### When do you use `chrome://tracing`?
```

When mysterious "glitches" happen. Profiling the app with the tracing tools

> gives you the insight on:

Reviewer: ```suggestion
gives you insights on:
```

* **Time slices spent by specific function calls** on different threads
* **Audio callback timing** in the timeline view

It usually shows the missed deadlines or the big garbage collection stops that
might cause unexpected audio glitches. This information is useful for triaging

> a bug and Chromium engineers will ask this if the local reproduction is not

Reviewer: ```suggestion
a bug and Chromium engineers will ask for tracing data if the local reproduction is not
```

feasible. General [instructions for tracing][chromium-tracing] are available
on chromium.org.

> ### When do you use the WebAudio devtool?

Reviewer: ```suggestion
### When do you use the **WebAudio** tab?
```

When you want to get a feel for how the application performs in the real world.
DevTools shows you a running estimate of render capacity, which indicates
how the web audio rendering engine is handling render tasks over a given
render budget (for example, approximately 2.67ms @ 48KHz). If the capacity
goes near 100%, that means your app is likely to produce glitches because the
renderer is failing to finish the work in the render budget.

## Use chrome://tracing

### How to capture tracing data

*The instructions below written are for Chrome 80 and later.*

For best results, close all other tabs and windows, and disable extensions.
Alternatively you can either [launch a new instance of Chrome][new-chrome-instance]
or use other builds from [different release channels][diff-channel] (e.g.
Beta or Canary). Once you have the browser ready, follow the steps below:

1. Open your application (web page) on a tab.
1. Open another tab and go to `chrome://tracing`.

> 1. Press **Record** button and select **Manually select settings**.

Reviewer: ```suggestion
1. Press the **Record** button and select **Manually select settings**.
```

1. Press the **None** buttons on both the **Record Categories** and
   **Disabled by Default Categories** sections.
1. In the **Record Categories** section, select the following:
    * `audio`
    * `blink_gc`
    * `media`
    * `v8.execute` (if you're interested in `AudioWorklet` JS code performance)
    * `webaudio`
1. In the **Disabled by Default Categories** section, select the following:

>     * `audio-worklet` (if you're interested in where `AudioWorklet` thread starts)

Reviewer: ```suggestion
    * `audio-worklet` (if you're interested in where the `AudioWorklet` thread starts)
```

    * `webaudio.audionode` (if you need the detailed trace for each `AudioNode`)

> 1. Press **Record** at the bottom.

Reviewer: ```suggestion
1. Press the **Record** button at the bottom.
```

1. Go back to your application tab and redo the steps that triggered the issue.
1. When you have enough trace data, go back to the tracing tab and press **Stop**.
1. The tracing tab will visualize the result (see below).
1. Press **Save** to save the tracing data.

<img src="1-hello-audio-worklet.jpg" class="w-screenshot w-screenshot-filled"
     alt="Screen shot after tracing has completed.">

### How to analyze tracing data

The tracing data visualizes how Chrome's web audio engine renders the audio.
The renderer has two different render modes: **Native mode** and
**Worklet mode**. Each mode uses a different threading model, so the tracing
results also differ.

> #### Native Mode

Reviewer: ```suggestion
#### Native mode
```

In Native mode, the [`AudioOutputDevice`][cr-audio-output-device] thread runs

> all the web audio codes. The `AudioOutputDevice` is a real-time priority thread

Reviewer: ```suggestion
all the web audio code. The `AudioOutputDevice` is a real-time priority thread
```

originating from the browser's Audio Service that is driven by the audio
hardware clock. If you see irregularity from the trace data in this lane,

> it means the callback timing from the device can be jittery. Linux/Pulse Audio
> combination is known for this problem. (See [crbug.com/825823](https://crbug.com/825823)
> and [crbug.com/864463](https://crbug.com/864463).)

Reviewer: ```suggestion
it means the callback timing from the device may be jittery. The combination
of Linux and Pulse Audio is known to have this problem. See the following Chromium issues
for more details: [#825823](https://crbug.com/825823), 
[#864463](https://crbug.com/864463).)
```

<img src="2e-box2d.jpg" class="w-screenshot w-screenshot-filled"
     alt="Screen shot of native mode tracing result." >

> #### Worklet Mode

Reviewer: ```suggestion
#### Worklet mode
```

In Worklet Mode, which is characterized by one thread jump from

> `AudioOutputDevice` to [`AudioWorklet` thread][cr-audio-worklet-thr], you

Reviewer: ```suggestion
`AudioOutputDevice` to the [`AudioWorklet` thread][cr-audio-worklet-thr], you
```

> should see well-aligned traces in two thread lanes as shown above.  When the

Reviewer: ```suggestion
should see well-aligned traces in two thread lanes as shown below.  When the
```

The common irregularity here is a big block caused by the garbage collection
or missed render deadlines. Both cases lead to glitches in the audio stream.

<img src="3e-hello-audio-worklet.png" class="w-screenshot w-screenshot-filled"
     alt="Screen shot of worklet mode tracing result.">

In both cases, the ideal tracing data is characterized by well-aligned audio
device callback invocations and render tasks being completed well within the
given render budget. The two screen shots above are great examples of the ideal
tracing data.

### Learning from real-world examples

#### Example 1: Render tasks going beyond render budget

The screenshot below ([crbug.com/796330](https://crbug.com/796330)) is a
typical example of when code in `AudioWorkletProcessor` takes too long and
goes beyond a given render budget. The callback timing is well behaved but
the audio processing function call of Web Audio API failed to complete the
work before the next device callback.

<img src="5e-render-budget.png" class="w-screenshot w-screenshot-filled"
     alt="Diagram showing audio glitch due to render task overflowing budget.">

**Your options:**

> * Reduce the workload of the audio graph. (Use fewer `AudioNodes`.)

Reviewer: ```suggestion
* Reduce the workload of the audio graph by using fewer `AudioNodes`.
```

Reviewer: @hoch is the literal interface name `AudioNodes`? Or is it `AudioNode`?

* Reduce the workload of your code in the `AudioWorkletProcessor`.
* Increase the base latency of `AudioContext`.

#### Example 2: Significant garbage collection on the worklet thread

Unlike on the native audio rendering thread, garbage collection is managed
on the worklet thread. That means if your code does memory allocation/deallocation
(e.g. new arrays) it eventually triggers a garbage collection which
synchronously blocks the thread. If the workload of web audio operations and
garbage collection is bigger than a given render budget, it results in
glitches in the audio stream. The screenshot is an extreme example of this
case.

> <img src="https://via.placeholder.com/250" class="w-screenshot w-screenshot-filled"

Author: @hoch - missing image.

Reviewer 2: Uploaded: 6e-garbage-collection.png

Reviewer: ```suggestion
<img src="6e-garbage-collection.png" class="w-screenshot w-screenshot-filled"
```

     alt="Screen shot showing garbage collection causing audio glitch.">

{% Aside %}
  Chrome's `AudioWorkletProcessor` implementation generates `Float32Array`s for
  the input and output buffer every audio processing callback, and this also
  slowly builds up the memory usage over time. The team has a plan to improve
  the design once the related specification is finalized.
{% endAside %}

**Your options:**

* Allocate the memory up front and reuse it whenever possible.
* Use different design patterns based on `SharedArrayBuffer`. Although this
  is not a perfect solution, several web audio apps use a similar pattern with
  `SharedArrayBuffer` to run the intensive audio code. See:
    - [Audio Worklet Design Pattern (Web Audio Power House)][webfu-audio-worklet-dp]
    - [Audio Device Client Prototype][aud-dev-client-proto]

> #### Example 3: Jittery audio device callback from AudioOutputDevice

Reviewer: ```suggestion
#### Example 3: Jittery audio device callback from `AudioOutputDevice`
```

The precise timing of audio callback is the most important thing for web audio.
This should be the most precise clock in your system. If the operating system
or its audio subsystem cannot guarantee a solid callback timing, all the
subsequent operations will be impacted. The image shown below is an example
of jittery audio callback. Compared to the previous two images, the interval
between each callback varies significantly.

<img src="4e-pulse-audio.png" class="w-screenshot w-screenshot-filled"
     alt="Screen shot comparing unstable vs stable callback timing.">

This is a known issue of Linux platform, which uses Pulse Audio as an audio
backend. This is still under investigation. (See
[crbug.com/825823](https://crbug.com/825823).)

> ## Use WebAudio DevTools

Reviewer: ```suggestion
## Use the **WebAudio** tab in Chrome DevTools
```

> You also can use the DevTools panel specifically designed for web audio. This

Reviewer: ```suggestion
You also can use the DevTools tab specifically designed for web audio. This
```

is less comprehensive compared to the tracing tool, but it is useful if you
want to gauge the running performance of your application.

Access the panel by clicking the "three dots" menu in the top right corner of
DevTools, then go to **More tools** then **WebAudio**.

> <img src="https://via.placeholder.com/250" class="w-screenshot w-screenshot-filled"

Author: @hoch - missing image.

Reviewer 2: Uploaded: 7e-devtools.png

     alt="Screen shot showing how to open WebAudio panel in Chrome DevTools.">

### WebAudio Panel

> <img src="https://via.placeholder.com/250" class="w-screenshot w-screenshot-filled"

Author: @hoch - missing image.

Reviewer 2: Uploaded: 8e-devtools.png

     alt="Screen shot of WebAudio panel in Chrome DevTools.">

This panel shows the information about running instances of `BaseAudioContext`.
Use it to see how the web audio renderer is performing on the page.

> Since a page can have multiple BaseAudioContexts, the context selector

Reviewer: ```suggestion
Since a page can have multiple `BaseAudioContext` instances, the context selector
```
